### DISCUSSION TITLE
How to attach to a program to execute gum functions?

### DISCUSSION THREAD

**Discussion by Xoffio** [frida/frida#2968] [Q&A]
I would like to attach to a pid but instead of loading an script I would like to execute gum functions. Is that possible?

When I attach to the program's pid i try running `gum_process_enumerate_ranges` and i get results but of the current program instead of the attached program.

This is what I have so far:
```rust
use std::ptr;

use frida::Frida;
use frida_gum_sys::{gboolean, gpointer, GumRangeDetails};
use lazy_static::lazy_static;

lazy_static! {
    static ref FRIDA: Frida = unsafe { Frida::obtain() };
}

unsafe extern "C" fn process_range(
    range: *const GumRangeDetails,
    _user_data: gpointer,
) -> gboolean {
    let r = &*range;

    println!(
        "range = {:?}, protection = {:?}, file = {:?}",
        r.range, r.protection, r.file
    );
    true as gboolean // Return true to continue enumeration
}

fn main() {
    let device_manager = frida::DeviceManager::obtain(&FRIDA);
    let local_device = device_manager.get_local_device().unwrap();
    let processes = local_device.enumerate_processes();

    let crackme_pid: Vec<&frida::Process> = processes
        .iter()
        .filter(|p| p.get_name() == "crackme-01")
        .collect();
    let crackme_pid = crackme_pid.first().unwrap().get_pid();
    let session = local_device.attach(crackme_pid).unwrap();

    let script_source = r#"
        console.log("test");
    "#;

    let mut script_option = frida::ScriptOption::default();
    let script = session
        .create_script(script_source, &mut script_option)
        .unwrap();

    script.load().unwrap();

    unsafe {
        frida_gum_sys::gum_process_enumerate_ranges(1 | 2, Some(process_range), ptr::null_mut());
    }

    script.unload().unwrap();
}
```

I really appreciate any help.

**Comment by s1341** [comment]
frida-gum can only be used in the current process. You need to build a shared-object which implements your desired frida-gum functionality, and inject it into the target program.

**Reply by Xoffio** [reply]
Hi @s1341! 
So I should create a lib like shown here -> [example.c](https://github.com/frida/frida-python/blob/72899a4315998289fb171149d62477ba7d1fcb91/examples/inject_library/example.c)
then inject it like this:

```rust
    let mut injector = Injector::new();
    injector
        .inject_library_file_sync(
            crackme_pid,
            "/tmp/example.dylib",
            "example_main",
            "hello world",
        )
        .expect("Could not inject library");
```
right?
What about the communication? is there any frida function I can use to handle the communication between these two? maybe the same way JavaScript is sending data back to frida-core?
I was trying to use gum functions to avoid any delays from JavaScript but now I am afraid introducing other type of delay doing it this way. 
Is there a better way to do this? 

Thank you for taking the time!

**Comment by s1341** [comment]
You should do everything inside the lib. Why do you need to communicate with an outside process?

**Reply by Xoffio** [reply]
Because I'm making an app with an UI so I need to get the data from the injected process and show it in the UI

**Comment by s1341** [comment]
Uh I see. You will need to come up with some kind of communications channel.

**Reply by oleavr** [reply]
Think we should expose our frida-pipe API -- it would make it easy to do this cross-platform.

**Reply by Xoffio** [reply]
@oleavr is frida-pipe what's being used to communicate frida-core with Js? What kind of communication is being used. If it's not too hard I could help with that

**Reply by Xoffio** [reply]
@oleavr is frida-pipe what's being used to communicate frida-core with Js? What kind of communication is being used? With some guidance I could work exposing the frida-pipe API.

**Reply by oleavr** [reply]
> @oleavr is frida-pipe what's being used to communicate frida-core with Js? What kind of communication is being used. If it's not too hard I could help with that

It is. The kind of communication depends on the platform -- we generally try to use the lowest level primitive to improve our odds in sandboxed scenarios. Shouldn't be too hard -- the main work is in doing bindings for the GIOStream API. The frida-python wrapper for it might be a useful source of inspiration, to get an idea of how little we expose.

**Reply by Xoffio** [reply]
Thank you. I will check the code tonight.

**Reply by Xoffio** [reply]
@oleavr  are these the ones that needs to be exposed https://github.com/frida/frida-core/tree/main/lib/pipe ?

**Reply by suidpit** [reply]
@Xoffio did you end up working on this? I am, too, looking at something similar, and it'd be really great to be able to implement a "fully fledged" instrumentation solution without leaving the gum land.

Right now, I am currently trying to adapt to a Javascript "middleware" where the actual instrumentation is done inside `CModule` interceptor callbacks, and javascript only handles communication between frida and the "control" code.

**Reply by Xoffio** [reply]
Hey @suidpit. I haven't done it. I started reading the code but never got near to expose the API. What I ended up doing was injecting a lib like I said here -> https://github.com/frida/frida/discussions/2968#discussioncomment-10670878 and then you can create your own way to communicate from your rust code to the injected lib

**Reply by suidpit** [reply]
I see, thanks for the quick response!
